pkg/aws/sts/cache.go

```diff
@@ -77,70 +81,82 @@ func DefaultCache(
+func (c *credentialsCache) evicted(key string, item interface{}) {
 	f := item.(*future.Future)
 	obj, err := f.Get(context.Background())
 
 	if err != nil {
-		log.WithField("pod.iam.role", role).Debugf("evicted credentials future had error: %s", err.Error())
+		log.WithField("cache.key", key).Debugf("evicted credentials future had error: %s", err.Error())
 		return
 	}
 
-	creds := obj.(*Credentials)
+	cachedCreds := obj.(*CachedCredentials)
 	select {
-	case c.expiring <- &RoleCredentials{Role: role, Credentials: creds}:
-		log.WithFields(CredentialsFields(creds, role)).Infof("notified credentials expire soon")
+	case c.expiring <- cachedCreds:
+		log.WithFields(CredentialsFields(cachedCreds.Identity, cachedCreds.Credentials)).Infof("notified credentials expire soon")
 		return
 	default:
 		return
 	}
 }
 
-func (c *credentialsCache) Expiring() chan *RoleCredentials {
+func (c *credentialsCache) Expiring() chan *CachedCredentials {
 	return c.expiring
 }
 
-func (c *credentialsCache) CredentialsForRole(ctx context.Context, role string) (*Credentials, error) {
-	logger := log.WithFields(log.Fields{"pod.iam.role": role})
-	item, found := c.cache.Get(role)
+func (c *credentialsCache) CredentialsForRole(ctx context.Context, identity *CredentialsIdentity) (*Credentials, error) {
+	logger := log.WithFields(log.Fields{"pod.iam.role": identity.Role})
+	key := c.cacheKey(identity)
+	item, found := c.cache.Get(key)
 
 	if found {
 		future, _ := item.(*future.Future)
 		val, err := future.Get(ctx)
 
 		if err != nil {
 			logger.Errorf("error retrieving credentials in cache from future: %s. will delete", err.Error())
-			c.cache.Delete(role)
+			c.cache.Delete(identity.Role)
 			return nil, err
 		}
 
 		cacheHit.Inc()
 
-		return val.(*Credentials), nil
+		cachedCreds := val.(*CachedCredentials)
+		return cachedCreds.Credentials, nil
 	}
 
 	cacheMiss.Inc()
 
 	issue := func() (interface{}, error) {
-		arn := c.arnResolver.Resolve(role)
+		arn := c.arnResolver.Resolve(identity.Role)
 		credentials, err := c.gateway.Issue(ctx, arn, c.sessionName, c.sessionDuration)
 		if err != nil {
 			errorIssuing.Inc()
 			logger.Errorf("error requesting credentials: %s", err.Error())
 			return nil, err
 		}
 
-		log.WithFields(CredentialsFields(credentials, role)).Infof("requested new credentials")
-		return credentials, err
+		cachedCreds := &CachedCredentials{
+			Identity:    identity,
+			Credentials: credentials,
+		}
+
+		log.WithFields(CredentialsFields(identity, credentials)).Infof("requested new credentials")
+		return cachedCreds, err
 	}
 	f := future.New(issue)
-	c.cache.Set(role, f, c.cacheTTL)
+	c.cache.Set(key, f, c.cacheTTL)
 
 	val, err := f.Get(ctx)
 	if err != nil {
-		c.cache.Delete(role)
+		c.cache.Delete(identity.Role)
 		return nil, err
 	}
 
-	return val.(*Credentials), nil
+	cachedCreds := val.(*CachedCredentials)
+	return cachedCreds.Credentials, nil
+}
+
+func (c *credentialsCache) cacheKey(identity *CredentialsIdentity) string {
```

Comment:
Thanks for adding. I'd change this so that perhaps `String()` on `CredentialsIdentity` generates the cache key, that way all the parts can be encapsulated away from `credentialsCache`. What do you think @stefansedich?

Reply:
That sounds entirely reasonable, just made the change.